File: 0261-graph-valid-tree/0261-graph-valid-tree.cpp
class Solution {
public:
    vector <int> ve[2001];
    int vis[2001];
    bool dfs(int s, int p){
        vis[s] = 1;
        bool res = true;
        for(auto x: ve[s]){
            if(x!=p){
                if(vis[x])return false;
                else {
                    res &= dfs(x,s);
                }
            }
        }
        return res;
    }
    bool validTree(int n, vector<vector<int>>& edges) {

        for(auto e: edges){
            ve[e[0]].push_back(e[1]);
            ve[e[1]].push_back(e[0]);
        }
        bool ok = dfs(0,-1);
        for(int i=0;i<n;i++){
            if(!vis[i])ok=false;
        }
        return ok;
    }
};

File: 0261-graph-valid-tree/0261-graph-valid-tree.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        
        preReq = {i:[] for i in range(n)}
        
        for a,b in edges:
            preReq[a].append(b)
            preReq[b].append(a)
            
        visit = set()
        
        def dfs(i,prev):
            
            if i in visit:
                return False
            
            visit.add(i)
            
            for nei in preReq[i]:
                if nei == prev:
                    continue
                    
                if not dfs(nei,i):
                    return False
                
            return True
        
        
        return dfs(0,-1) and n == len(visit)